udmis/src/main/java/com/google/bos/udmi/monitoring/LogTimeSeries.java

```diff
@@ -0,0 +1,179 @@
+          // associated with a Message body, which is how it is used elsewhere.
+          monitoring.metric.envelope = new Envelope();
+          monitoring.metric.envelope.subFolder = SubFolder.MONITORING;
+          loadMetricFields(monitoring.metric, log);
+          output.emitMetric(monitoring);
+        }
+      }
+    }
+  }
+
+  private int findSameLog(long k, LogTailEntry log) {
+    // TODO: Revise when output storage requirements are known.
+    LinkedList<LogTailEntry> logs = this.get(k);
+    for (int i = 0; i < logs.size(); i++) {
+      if (log.severity.equals(Severity.ERROR) && logs.get(i).serviceName.equals(
+          log.severity.toString())) {
+        if (logs.get(i).statusMessage.equals(log.statusMessage)) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  /*
+   * Generate the bucket key for a LogTailEntry.
+   */
+  private long keyFor(LogTailEntry log) {
+    long t = log.timestamp.toEpochMilli();
+    t = t - (t % (LOG_ENTRIES_BUCKET_SECONDS * 1000));
+    return t;
+  }
+
+  private void loadMetricFields(MonitoringMetric metric, LogTailEntry log) throws IOException {
+    metric.event_system.timestamp = Date.from(log.timestamp);
+    metric.event_system.version = "1.4.1"; // TODO: Where to get this value
+
+    udmi.schema.Entry entry = new udmi.schema.Entry();
+    entry.level = severityToLogSeverity(log.severity).getNumber();
+    entry.message = log.statusMessage;
+
+    if (!log.resourceName.isEmpty()) {
+      loadMetricFieldsFromResourceName(metric, log);
+    }
+
+    if (!log.statusMessage.isEmpty()) {
+      entry.message = log.statusMessage;
+      if (log.severity.toString().equals("ERROR")) {
+        Pattern pattern = Pattern.compile("Device [`'](\\d+)[`'] is ");
+        Matcher matcher = pattern.matcher(log.statusMessage);
+        if (matcher.find()) {
+          metric.envelope.deviceNumId = matcher.group(1);
+        }
+      }
+    }
+
+    metric.event_system.logentries.add(entry);
+  }
+
+  private void loadMetricFieldsFromResourceName(MonitoringMetric metric, LogTailEntry log) {
+    // Example:
+    // projects/essential-keep-197822/locations/us-central1/...
+    // registries/UDMS-REFLECT/devices/IN-HYD-SAR2
+    Pattern pattern = Pattern.compile(
+        "projects/(?<project>[^/]+)/locations/(?<location>[^/]+)"
+            + "/registries/(?<registry>[^/]+)/devices/(?<device>[^/]+)");
+    Matcher matcher = pattern.matcher(log.resourceName);
+    if (matcher.find()) {
+      metric.envelope.projectId = matcher.group("project");
+      metric.envelope.deviceRegistryLocation = matcher.group("location");
+      metric.envelope.deviceRegistryId = matcher.group("registry");
+      metric.envelope.deviceId = matcher.group("device");
+    } // else do something?
+  }
+
+  private LogSeverity severityToLogSeverity(Severity severity) throws IOException {
+    // Subtle: These are two different types with different enum value ranges. Translate manually.
+    switch (severity) {
```

Comment:
(other common ways to do this are with an (immutable) map or with a string conversion) -- just FYI, no need to change